### PR TITLE
Khalil/ 1725 authenticate libp2p peer before decoding message payload

### PR DESCRIPTION
**Pull request by kc1116**
This PR adds a check if a peer is a staked node in the TopicValidator before decoding the message payload. This is done by providing the libP2P node top-level TopicValidator with a **_isStaked_** callback func from the middleware whenever a channel is subscribed to. For public channels, we negate this check by returning a default callback that will return true for all nodes.

### CODE REVIEW

File: network/p2p/topic_validator_test.go [hunk 1]
```diff
@@ -23,17 +23,91 @@ import (
...
+	}
+
+	// node1 is connected to node2
+	// sn1 <-> sn2
+	require.NoError(t, sn1.AddPeer(context.TODO(), *host.InfoFromHost(sn2.Host())))
+
+	// sn1 will subscribe with is staked callback that should force the TopicValidator to drop the message received from sn2
+	sub1, err := sn1.Subscribe(topic, unittest.NetworkCodec(), isStaked)
+	require.NoError(t, err)
+	_, err = sn2.Subscribe(topic, unittest.NetworkCodec(), unittest.IsStakedTopicValidatorFunc())
```
Comment (peterargue) [inline network/p2p/topic_validator_test.go:69]:
```suggestion
	// sn2 will subscribe with an unauthenticated callback to allow it to send the unauthenticated message
	_, err = sn2.Subscribe(topic, unittest.NetworkCodec(), unittest.IsStakedTopicValidatorFunc())
```

File: network/p2p/topic_validator_test.go [hunk 2]
```diff
@@ -23,17 +23,91 @@ import (
 	"github.com/onflow/flow-go/utils/unittest"
 )
 
+// TestTopicValidator_Unstaked tests that the libP2P node topic validator rejects messages from unstaked nodes
+func TestTopicValidator_Unstaked(t *testing.T) {
```
Comment (peterargue) [inline network/p2p/topic_validator_test.go:27]: Can you also add a test for an unauthenticated subscription

File: utils/unittest/fixtures.go [hunk 3]
```diff
@@ -2080,3 +2081,9 @@ func TransactionResultsFixture(n int) []flow.TransactionResult {
 	}
 	return results
 }
+
+func IsStakedTopicValidatorFunc() func(peer.ID) bool {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply utils/unittest/fixtures.go]: https://github.com/onflow/flow-go/pull/2667/commits/7389ef661fb79bc65f70f86bd95a0910efe53501

File: network/p2p/libp2pNode.go [hunk 4]
```diff
@@ -172,7 +172,7 @@ func (n *Node) ListPeers(topic string) []peer.ID {
 // Subscribe subscribes the node to the given topic and returns the subscription
 // Currently only one subscriber is allowed per topic.
 // NOTE: A node will receive its own published messages.
-func (n *Node) Subscribe(topic flownet.Topic, codec flownet.Codec, validators ...validator.MessageValidator) (*pubsub.Subscription, error) {
+func (n *Node) Subscribe(topic flownet.Topic, codec flownet.Codec, isStaked isStakedFunc, validators ...validator.MessageValidator) (*pubsub.Subscription, error) {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/libp2pNode.go]: https://github.com/onflow/flow-go/pull/2667/commits/7389ef661fb79bc65f70f86bd95a0910efe53501

File: network/p2p/middleware.go [hunk 5]
```diff
@@ -507,16 +517,25 @@ func (m *Middleware) Subscribe(channel network.Channel) error {
 
 	topic := network.TopicFromChannel(channel, m.rootBlockID)
 
+	// NOTE: for public channels the callback used to check if a node is staked will
+	// return true for every node.
+	isStakedFunc := defaultIsStakedFunc
 	var validators []psValidator.MessageValidator
 	if !network.PublicChannels().Contains(channel) {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/middleware.go]: https://github.com/onflow/flow-go/pull/2667/commits/7389ef661fb79bc65f70f86bd95a0910efe53501

File: network/p2p/middleware.go [hunk 6]
```diff
@@ -507,16 +517,27 @@ func (m *Middleware) Subscribe(channel network.Channel) error {
...
 			psValidator.AuthorizedSenderValidator(m.log, channel, m.ov.Identity),
 		)
+
+		// NOTE: For non-public channels the libP2P node topic validator will reject
+		// messages from unstaked nodes.
+		authenticatePeer = func(id peer.ID) bool {
+			_, ok := m.ov.Identity(id)
+			return ok
+		}
+	} else {
```
Comment (peterargue) [inline network/p2p/middleware.go:534]: nit: the logic "if not, else" can be confusing and require reading the code a couple times. better to flip the order so conditional is more intuitive

File: network/p2p/topic_validator_test.go [hunk 7]
```diff
@@ -86,65 +83,102 @@ func TestAuthorizedSenderValidator_Unauthorized(t *testing.T) {
...
+	checkReceive(timedCtx, t, data1, sub1, nil, true)
+
+	// sn2 also gets the message (as part of the libp2p loopback of published topic messages)
+	checkReceive(timedCtx, t, data1, sub2, nil, true)
 
 	unittest.RequireReturnsBefore(t, wg.Wait, 5*time.Second, "could not receive message on time")
 
-	// expecting 1 warn calls for each rejected message from unauthorized node
-	require.Equalf(t, uint64(1), hookCalls, "expected 1 warning to be logged")
+	// expecting 1 warn calls for each rejected message from ejected node
```
Comment (peterargue) [inline network/p2p/topic_validator_test.go:162]:
```suggestion
	// expecting no warn calls for rejected messages
```

File: network/p2p/topic_validator_test.go [hunk 8]
```diff
@@ -86,65 +83,102 @@ func TestAuthorizedSenderValidator_Unauthorized(t *testing.T) {
...
-	checkReceive(timedCtx, t, data2, sub3, nil, true)
+	// assert that the nodes are connected as expected
+	require.Eventually(t, func() bool {
+		return len(sn1.ListPeers(topic.String())) > 0 &&
+			len(sn2.ListPeers(topic.String())) > 0
+	}, 3*time.Second, 100*time.Millisecond)
+
+	timedCtx, cancel5s := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel5s()
+	// create a dummy block proposal to publish from our SN node
```
Comment (peterargue) [inline network/p2p/topic_validator_test.go:142]:
```suggestion
	// create a dummy sync request to publish from our SN node
```

File: network/p2p/topic_validator_test.go [hunk 9]
```diff
@@ -86,65 +83,102 @@ func TestAuthorizedSenderValidator_Unauthorized(t *testing.T) {
...
 
-	// an1 also gets the message
-	checkReceive(timedCtx, t, data1, sub3, nil, true)
+	unittest.RequireReturnsBefore(t, wg.Wait, 5*time.Second, "could not receive message on time")
 
-	timedCtx, cancel2s := context.WithTimeout(context.Background(), 2*time.Second)
-	defer cancel2s()
-	header = unittest.BlockHeaderFixture()
-	data2 := getMsgFixtureBz(t, &messages.BlockProposal{Header: &header})
+	// expecting 1 warn calls for each rejected message from ejected node
```
Comment (peterargue) [inline network/p2p/topic_validator_test.go:98]:
```suggestion
	// expecting 1 warn calls for each rejected message from unauthenticated node
```

File: network/p2p/middleware.go [hunk 10]
```diff
@@ -58,7 +58,17 @@ const (
...
 )
 
-var _ network.Middleware = (*Middleware)(nil)
+var (
+	_ network.Middleware = (*Middleware)(nil)
+
+	// unauthenticated is the default callback func used to check if a node
+	// is staked or not in the topic validator before decoding network messages.
+	// This default func should only be used on public network channels, since all
+	// nodes can participate on public channels this func always returns true.
```
Comment (peterargue) [inline network/p2p/middleware.go:67]: let's make this warning louder
```suggestion
	// unauthenticated should ONLY be used for public network channels to allow
	// messages from any peer. This is NOT safe for channels on the staked network
	// as it skips authenticating the peer which may result in resource exhaustion attacks.
```

**Review by peterargue — APPROVED** [review]
looks good. added a few small suggestions/nits

File: network/p2p/libp2pNode.go [hunk 11]
```diff
@@ -181,7 +181,7 @@ func (n *Node) Subscribe(topic flownet.Topic, codec flownet.Codec, validators ..
 	tp, found := n.topics[topic]
 	var err error
 	if !found {
-		topicValidator := validator.TopicValidator(n.logger, codec, validators...)
+		topicValidator := validator.TopicValidator(n.logger, codec, authenticatePeer, validators...)
```
Comment (smnzhu) [inline network/p2p/libp2pNode.go:184]:
```suggestion
		topicValidator := validator.TopicValidator(n.logger, codec, authorizePeer, validators...)
```

File: network/p2p/middleware.go [hunk 12]
```diff
@@ -58,7 +58,17 @@ const (
...
+	unauthenticated = func(_ peer.ID) bool { return true }
```
Comment (smnzhu) [inline network/p2p/middleware.go:68]: The naming is confusing here. We should call it `allAuthorized` or something.

File: network/p2p/middleware.go [hunk 13]
```diff
@@ -58,7 +58,17 @@ const (
...
+)
+
+type validatorPeerAuthenticationFunc func(id peer.ID) bool
```
Comment (smnzhu) [inline network/p2p/middleware.go:71]:
```suggestion
type validatorPeerAuthorizationFunc func(id peer.ID) bool
```

File: utils/unittest/fixtures.go [hunk 14]
```diff
@@ -2080,3 +2081,9 @@ func TransactionResultsFixture(n int) []flow.TransactionResult {
 	}
 	return results
 }
+
+func UnauthenticatedTopicValidatorFunc() func(peer.ID) bool {
```
Comment (smnzhu) [inline utils/unittest/fixtures.go:2085]: similar comment here

File: network/p2p/libp2pNode.go [hunk 15]
```diff
@@ -172,7 +172,7 @@ func (n *Node) ListPeers(topic string) []peer.ID {
 // Subscribe subscribes the node to the given topic and returns the subscription
 // Currently only one subscriber is allowed per topic.
 // NOTE: A node will receive its own published messages.
-func (n *Node) Subscribe(topic flownet.Topic, codec flownet.Codec, validators ...validator.MessageValidator) (*pubsub.Subscription, error) {
+func (n *Node) Subscribe(topic flownet.Topic, codec flownet.Codec, authenticatePeer validatorPeerAuthenticationFunc, validators ...validator.MessageValidator) (*pubsub.Subscription, error) {
```
[the remark this answers is not in the document]
Comment (peterargue) [reply network/p2p/libp2pNode.go:175]: I think authenticate is accurate. we're checking that the peer ID from the message signing key matches an id in our peer list. i.e. checking that the peer is who they say they are and that we know them

File: network/validator/pubsub/validator.go [hunk 16]
```diff
@@ -64,7 +64,11 @@ type ValidatorData struct {
 	From              peer.ID
 }
 
-func TopicValidator(logger zerolog.Logger, codec network.Codec, validators ...MessageValidator) pubsub.ValidatorEx {
+func TopicValidator(log zerolog.Logger, codec network.Codec, authenticatePeer func(peer.ID) bool, validators ...MessageValidator) pubsub.ValidatorEx {
```
Comment (smnzhu) [inline network/validator/pubsub/validator.go:67]:
```suggestion
func TopicValidator(log zerolog.Logger, codec network.Codec, peerFilter func(peer.ID) bool, validators ...MessageValidator) pubsub.ValidatorEx {
```

File: utils/unittest/fixtures.go (hunk 14 above)
Comment (smnzhu) [inline utils/unittest/fixtures.go:2085]:
```suggestion
func AllowAllPeerFilter() func(peer.ID) bool {
```

File: network/p2p/middleware.go (hunk 12 above)
Comment (smnzhu) [inline network/p2p/middleware.go:68]:
```suggestion
	allowAll = func(_ peer.ID) bool { return true }
```

**Comment by AlexHentschel** [comment]
what does this command do: https://github.com/onflow/flow-go/blob/9807c9010b3230418f49464cb2632af84547ade2/network/p2p/libp2pNode.go#L180
I am confused because we entirely ignore the return value from `GetTopics()`

**Comment by kc1116** [comment]
@AlexHentschel 
> what does this command do: https://github.com/onflow/flow-go/blob/9807c9010b3230418f49464cb2632af84547ade2/network/p2p/libp2pNode.go#L180
I am confused because we entirely ignore the return value from `GetTopics()`

via @yhassanzadeh13 it flushes the subscription requests that have already been cached into the list of topics the node subscribed to.

**Review by peterargue — APPROVED** [review]

**Comment by yhassanzadeh13** [comment]
> @AlexHentschel
> 
> > what does this command do: https://github.com/onflow/flow-go/blob/9807c9010b3230418f49464cb2632af84547ade2/network/p2p/libp2pNode.go#L180
> > 
> > I am confused because we entirely ignore the return value from `GetTopics()`
> 
> via @yhassanzadeh13 it flushes the subscription requests that have already been cached into the list of topics the node subscribed to.


I looked further into the libp2p code and it seems like we no longer need to have this invocation. I will do remove this in a separate PR just in case we will be able to revert it back if needed. 
 https://github.com/onflow/flow-go/blob/9807c9010b3230418f49464cb2632af84547ade2/network/p2p/libp2pNode.go#L180
@AlexHentschel @kc1116

**Review by AlexHentschel — COMMENTED** [review]
Comments for `BFT Engineering and PR Open house` (Friday June 24th)

**Review by smnzhu — APPROVED** [review]

**Pull request merged**